File: cyclespace.cpp
#include "cyclespace.h"
#include "mstprim.h"

#include <algorithm>
#include <map>
#include <queue>
#include <set>

CycleSpaceResult CycleSpace::solve(const std::vector<std::vector<int>>& adj,
                                    const std::vector<std::vector<int>>& weight)
{
    CycleSpaceResult res;
    const int n = static_cast<int>(adj.size());
    if (n <= 0) return res;

    // Build MST using Prim (input already undirected — caller's responsibility)
    MSTResult mst = MSTPrim::solve(adj, weight);
    if (!mst.connected) return res;

    // Build MST edge set for fast lookup
    std::set<std::pair<int,int>> mstEdgeSet;
    for (auto& e : mst.edges)
        mstEdgeSet.insert({std::min(e.first, e.second), std::max(e.first, e.second)});

    // Build MST adjacency lists for BFS path finding
    std::vector<std::vector<int>> mstAdj(n);
    for (auto& e : mst.edges) {
        mstAdj[e.first].push_back(e.second);
        mstAdj[e.second].push_back(e.first);
    }

    // Find non-tree edges
    for (int i = 0; i < n; ++i) {
        for (int j = i + 1; j < n; ++j) {
            if (!adj[i][j]) continue;
            auto key = std::make_pair(i, j);
            if (mstEdgeSet.count(key)) continue;  // tree edge

            // BFS in MST from i to j to find the tree path
            std::vector<int> parent(n, -1);
            std::vector<bool> visited(n, false);
            std::queue<int> q;
            q.push(i);
            visited[i] = true;
            while (!q.empty() && !visited[j]) {
                int v = q.front(); q.pop();
                for (int u : mstAdj[v]) {
                    if (!visited[u]) {
                        visited[u] = true;
                        parent[u] = v;
                        q.push(u);
                    }
                }
            }

            // Reconstruct path from i to j
            std::vector<int> path;
            for (int v = j; v != -1; v = parent[v])
                path.push_back(v);
            std::reverse(path.begin(), path.end());

            // Build cycle: close the loop
            FundamentalCycle cycle;
            cycle.vertices = path;
            cycle.vertices.push_back(path[0]);
            cycle.cotreeEdge = {i, j};

            // Collect edges: tree path edges + the non-tree closing edge
            for (int k = 0; k + 1 < static_cast<int>(path.size()); ++k) {
                int a = path[k], b = path[k + 1];
                cycle.edges.push_back({std::min(a, b), std::max(a, b)});
            }
            cycle.edges.push_back({std::min(i, j), std::max(i, j)});

            res.cycles.push_back(cycle);
        }
    }

    return res;
}

std::vector<std::pair<int,int>>
CycleSpace::symmetricDifference(const CycleSpaceResult& result,
                                 const std::vector<int>& selectedIndices)
{
    std::map<std::pair<int,int>, int> freq;
    for (int idx : selectedIndices) {
        if (idx < 0 || idx >= static_cast<int>(result.cycles.size())) continue;
        for (auto& e : result.cycles[idx].edges)
            freq[e]++;
    }

    std::vector<std::pair<int,int>> symDiff;
    for (auto& [e, cnt] : freq)
        if (cnt % 2 == 1) symDiff.push_back(e);

    std::sort(symDiff.begin(), symDiff.end());
    return symDiff;
}

File: web/js/cyclespace.js
"use strict";

// Fundamental cycle system based on MST (Prim's algorithm from Lab 4).
// Both methods expect already-symmetrized (undirected) input — caller's responsibility.

window.CycleSpace = {

    solve(adjMatrix, weightMatrix) {
        const n = adjMatrix.length;
        const result = { cycles: [] };
        if (n <= 0) return result;

        // Build MST via Prim (reuse Lab 4 module)
        const mst = MSTPrim.solve(adjMatrix, weightMatrix);
        if (!mst.connected) return result;

        // MST edge set for fast lookup: keys are "min,max"
        const mstEdgeSet = new Set();
        for (const [u, v] of mst.edges)
            mstEdgeSet.add(Math.min(u, v) + ',' + Math.max(u, v));

        // MST adjacency lists for BFS
        const mstAdj = Array.from({ length: n }, () => []);
        for (const [u, v] of mst.edges) {
            mstAdj[u].push(v);
            mstAdj[v].push(u);
        }

        // Find non-tree edges and build one fundamental cycle per edge
        for (let i = 0; i < n; i++) {
            for (let j = i + 1; j < n; j++) {
                if (!adjMatrix[i][j]) continue;
                if (mstEdgeSet.has(i + ',' + j)) continue;  // tree edge

                // BFS in MST from i to j
                const parent = new Array(n).fill(-1);
                const visited = new Array(n).fill(false);
                const queue = [i];
                visited[i] = true;
                while (queue.length > 0 && !visited[j]) {
                    const v = queue.shift();
                    for (const u of mstAdj[v]) {
                        if (!visited[u]) {
                            visited[u] = true;
                            parent[u] = v;
                            queue.push(u);
                        }
                    }
                }

                // Reconstruct path i → j
                const path = [];
                for (let v = j; v !== -1; v = parent[v]) path.push(v);
                path.reverse();

                // Build cycle: close the loop
                const vertices = [...path, path[0]];

                // Collect edges: tree path edges + non-tree closing edge
                const edges = [];
                for (let k = 0; k + 1 < path.length; k++) {
                    const a = path[k], b = path[k + 1];
                    edges.push([Math.min(a, b), Math.max(a, b)]);
                }
                edges.push([Math.min(i, j), Math.max(i, j)]);

                result.cycles.push({ vertices, edges, cotreeEdge: [i, j] });
            }
        }

        return result;
    },

    symmetricDifference(result, selectedIndices) {
        // Count how many selected cycles contain each edge
        const freq = new Map();
        for (const idx of selectedIndices) {
            if (idx < 0 || idx >= result.cycles.length) continue;
            for (const [u, v] of result.cycles[idx].edges) {
                const key = u + ',' + v;
                freq.set(key, (freq.get(key) || 0) + 1);
            }
        }

        // Keep edges with odd frequency
        const symDiff = [];
        for (const [key, cnt] of freq) {
            if (cnt % 2 === 1) {
                const parts = key.split(',');
                symDiff.push([parseInt(parts[0], 10), parseInt(parts[1], 10)]);
            }
        }

        symDiff.sort((a, b) => a[0] !== b[0] ? a[0] - b[0] : a[1] - b[1]);
        return symDiff;
    }
};
